libsolidity/interface/CompilerStack.cpp

```diff
@@ -695,15 +695,15 @@ vector<string> CompilerStack::contractNames() const
 	return contractNames;
 }
 
-string const CompilerStack::lastContractName() const
+string const CompilerStack::lastContractName(optional<string> const& _sourceName) const
 {
 	if (m_stackState < AnalysisPerformed)
 		BOOST_THROW_EXCEPTION(CompilerError() << errinfo_comment("Parsing was not successful."));
 	// try to find some user-supplied contract
 	string contractName;
 	for (auto const& it: m_sources)
-		for (ASTPointer<ASTNode> const& node: it.second.ast->nodes())
-			if (auto contract = dynamic_cast<ContractDefinition const*>(node.get()))
+		if (_sourceName.value_or(it.first) == it.first)
```

Comment:
Hah, does C++ copies Rust's `unwrap_or` with` value_or`, or vice versa?

Only as a remark: In any case not sure this is all that readable, took me a second to figure out.

Reply:
I don't understand this.